### PR TITLE
vigo/auraWorks

### PR DESCRIPTION
**Pull request by vigo2**
[core] speed up aura handling by caching the minimum expires time, to shortcut most advance() calls; active auras again only store auras that may expire; advance() uses a faster iteration scheme

I'll sprinkle some comments below.

### CODE REVIEW

File: sim/core/aura.go [hunk 1]
```diff
@@ -203,14 +204,16 @@ type auraTracker struct {
 
 	aurasByTag map[string][]*Aura
 
-	// IDs of Auras that are active, in no particular order.
+	// IDs of Auras that may expire and are currently active, in no particular order.
 	activeAuras []*Aura
 
+	// caches the minimum expires time of all active auras; reset to 0 on Activate(), Deactivate(), and Refresh()
+	minExpires time.Duration
```
Comment (vigo2) [inline sim/core/aura.go:211]: Since Activate() calls Refresh(), this resetting is only implicit.

File: sim/core/aura.go [hunk 2]
```diff
@@ -203,14 +204,16 @@ type auraTracker struct {
...
+
 	// Auras that have a non-nil XXX function set and are currently active.
 	onCastCompleteAuras   []*Aura
 	onSpellHitAuras       []*Aura
 	onPeriodicDamageAuras []*Aura
-	onMeleeAttackAuras    []*Aura
```
Comment (vigo2) [inline sim/core/aura.go:213]: Unused, as far as I can see.

File: sim/core/aura.go [hunk 3]
```diff
@@ -286,24 +288,16 @@ func (unit *Unit) GetOrRegisterAura(aura *Aura) *Aura {
 }
 
 func (at *auraTracker) GetAurasWithTag(tag string) []*Aura {
-	if auras, ok := at.aurasByTag[tag]; ok {
-		return auras
-	} else {
-		return []*Aura{}
-	}
+	return at.aurasByTag[tag]
```
Comment (vigo2) [inline sim/core/aura.go:291]: Returning the nil slice is more shorter and more idiomatic.

File: sim/core/aura.go [hunk 4]
```diff
@@ -286,24 +288,16 @@ func (unit *Unit) GetOrRegisterAura(aura *Aura) *Aura {
...
+		if aura.active {
+			return true
 		}
 	}
-	return active
-}
-func (at *auraTracker) HasActiveAuraWithTag(tag string) bool {
-	return len(at.GetActiveAurasWithTag(tag)) > 0
+	return false
 }
```
Comment (vigo2) [inline sim/core/aura.go:301]: GetActiveAurasWithTag() was only used for HasActiveAuraWithTag(), so I cut the middleman. I don't think the tag support is really needed, but it's probably easier to use.

File: sim/core/aura.go [hunk 5]
```diff
@@ -386,16 +379,22 @@ func (at *auraTracker) reset(sim *Simulation) {
...
+		if aura.expires <= sim.CurrentTime {
+			aura.Deactivate(sim)
+			goto restart // activeAuras have changed
+		}
+		if aura.expires < minExpires {
+			minExpires = aura.expires
 		}
 	}
+	at.minExpires = minExpires
 }
```
Comment (vigo2) [inline sim/core/aura.go:398]: Using "goto" since I think it's easier to understand than recursion, or adding an outer loop. Iterating in order is quite a bit faster than a reverse iteration, and in 99% of all cases (at least for the rogue sim I instrumented) nothing is deactivated, so the extra recursion(s) don't matter.

**Comment by lologarithm** [comment]
I am a huge fan of the min aura time improving perf!!!

File: sim/core/aura.go (hunk 2 above)
Comment (lologarithm) [reply sim/core/aura.go:213]: I know this used to be used but there has been a lot of aura refactoring recently so it might be now.

File: sim/core/aura.go (hunk 3 above)
Comment (lologarithm) [reply sim/core/aura.go:291]: :+1:

File: sim/core/aura.go (hunk 4 above)
Comment (jimmyt857) [reply sim/core/aura.go:301]: If you have another idea for how to manage groups of auras, a PR would be welcome.

File: sim/core/aura.go (hunk 5 above)
Comment (jimmyt857) [reply sim/core/aura.go:398]: FWIW, the case that forced me to add the extra check was Badge of the Swarmguard. The expiration of the active aura triggers the expiration of the stacks aura, so both expire at once.

File: sim/mage/TestFrost.results [hunk 6]
```diff
@@ -455,7 +455,7 @@ dps_results: {
 dps_results: {
  key: "TestFrost-AllItems-RobeoftheElderScribes-28602"
  value: {
-  dps: 1656.125661967007
+  dps: 1653.905027551007
  }
```
Comment (jimmyt857) [inline sim/mage/TestFrost.results:459]: Any hunch as to why this is changing some test results? I'm concerned there's some edge case where auras are not getting cleaned up when they should, due to the minExpires check.

Comment (vigo2) [reply sim/mage/TestFrost.results:459]: The changes are due to the iteration order change, and didn't change afterwards: for rogues and warriors I suspect mongoose procs might be affected, since MH and OH share the same RandomFloat-label, if I'm not mistaken.

Comment (jimmyt857) [reply sim/mage/TestFrost.results:459]: Ok that's fine then.

Comment (vigo2) [reply sim/mage/TestFrost.results:459]: FWIW, I've validated that the set of auras to expire or keep at the start of advance() were actually expired or kept afterwards. That only broke for the Adrenaline Rush expires -> extra tick -> extra special attack -> mongoose proc chain, where (correctly though) an extra aura appears.

File: sim/core/aura.go (hunk 4 above)
Comment (vigo2) [reply sim/core/aura.go:301]: If tagging is only used for inter-aura priority management (IEA overwriting SA etc.), I think that could be handled like before. I haven't looked into other use cases, though, and grouping might help in other scenarios.
